### PR TITLE
Suricata

### LINKED ISSUE #1045: Bundle Suricata on Linux

### LINKED ISSUE #1046: Bundle Suricata on Windows

### LINKED ISSUE #1047: Bundle Suricata on macOS

### LINKED ISSUE #1206: Enable Suricata

### PR DESCRIPTION
**Pull request by henridf**
This PR adds support for Suricata to Brim. 

Closes #1206 
Closes #1045 (but the packages fail due to to #1216)
Closes #1046 
Closes #1047

### CODE REVIEW

**Comment by jameskerr** [comment]
The front end code looks complete. Thank you! I did a proper `nom install` and tried running this branch, but got this error in the console.

```
flag provided but not defined: -suricataupdater
unknown flag: "-l localhost:9867 -data /Users/jkerr/work/brim/run/data/spaces -config /Users/jkerr/work/brim/run/zqd-config.yaml -suricatarunner /Users/jkerr/work/brim/zdeps/suricata/suricatarunner -suricataupdater /Users/jkerr/work/brim/zdeps/suricata/suricataupdater -zeekrunner /Users/jkerr/work/brim/zdeps/zeek/zeekrunner -brimfd=3"
```

Is that expected in this PR?

**Comment by philrz** [comment]
@jameskerr: It looks like maybe @henridf might have been testing with a manually-swapped-in `zqd`, as I was also able to repro what you described by using the branch as-is with its current `zq` pointer in `package.json`. However, if I first did:

```
npm install https://github.com/brimsec/zq#71daeec268b24d047f6d7b1ba0c08b2137e8fb50
```

...before my `npm install` to get a more recent `zq` commit, now I was able to import a pcap and see alert events, e.g.:

image

Now that I can see the events, the next glitch I notice is that the `community_id` fields all appear to be blank, even though the Suricata config YAML makes it look like they should be generated. I'll bet @henridf could figure that out when his day starts again, but now I'm curious, so I may dig a little more.

**Comment by philrz** [comment]
Here's a progress report. Concerned that maybe Community ID was flat-out busted in Suricata, I did a compile from source of "regular" Suricata v5.0.3 on my Mac. Of course, I was missing some of the needed dependencies, so I used the ones from @henridf's [macos-build.yaml](https://github.com/brimsec/build-suricata/blob/master/.github/workflows/macos-build.yaml) (leaving off `pyinstaller` since I knew I'd not need it.) Hence:

```
        brew install rust pkg-config
        brew install jansson libmagic libnet libyaml lz4 nspr nss pcre bzip2
        brew install autoconf automake libtool
        pip3 install pyyaml
```

Some of these I already had installed, and some I didn't. In any case, that brought me to the point where Suricata compiled/installed/ran ok. At that point I ran my pcap through it and saw the Community ID fields were being populated ok. Then I went back to my Brim-installed Suricata to start looking at what the differences might be. I ran another pcap through it just for kicks, and now the Community ID field was populated ok in Brim as well!

Conclusion: I'm assuming that one of those dependencies I needed to compile Suricata is also needed at runtime for the Community ID to be generated. However, I can't guess which one without looking at the code, which is not my area of expertise.

**Comment by philrz** [comment]
Here's a final update for today. I fiddled on a scratch Actions host by installing dependencies one at time, through which I deduced it was the `brew install nss` that made all the difference. I then looked at `brew ls --verbose nss` to see what it had installed, and with that output I learned that deleting all of `/usr/local/Cellar/nss/3.58/lib/*.dylib` made Community ID stop working. I then tried messing around with removing individual `.dylib` files to see if just one of them made all the difference, but those tests indicated more than one (but not all of them) were needed. For quick reference, the contents on a pristine install:

```
$ ls -l /usr/local/Cellar/nss/3.58/lib/*.dylib
-r--r--r--  1 runner  staff   959176 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libfreebl3.dylib
-r--r--r--  1 runner  staff  1196944 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libnss3.dylib
-r--r--r--  1 runner  staff   132496 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libnssckbi-testlib.dylib
-r--r--r--  1 runner  staff   476304 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libnssckbi.dylib
-r--r--r--  1 runner  staff   192936 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libnssdbm3.dylib
-r--r--r--  1 runner  staff   186488 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libnssutil3.dylib
-r--r--r--  1 runner  staff    53712 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libpkcs11testmodule.dylib
-r--r--r--  1 runner  staff   197320 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libsmime3.dylib
-r--r--r--  1 runner  staff   327400 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libsoftokn3.dylib
-r--r--r--  1 runner  staff   339952 Nov 17 03:04 /usr/local/Cellar/nss/3.58/lib/libssl3.dylib
```

Looking at `build-suricata/Makefile-macOS.brim`, it looks like there's missing `.a` equivalent entries for some of these `.dylib` entries, so I assume some more static linking is in order. I'll let @henridf take it from here.

**Comment by henridf** [comment]
@philrz @jameskerr I fixed the zq version. Tests are now passing except the Ubuntu CI which has failures which _seem_ unrelated but I'm not sure. @jameskerr any thoughts on whether it's related?

**Comment by henridf** [comment]
Thanks @philrz for the sleuthing on the missing libs! That will be fixed in the build-suricata and zq repos, so I think this PR can continue on its way.

**Comment by jameskerr** [comment]
@henridf the integration test failures do look related to the suricata changes, although it just might be some strange GitHub actions problem.

Here is the screenshot that was taken after a few tests failed:

failure-c58c80b6

Here was zqd log output from the first failed ubuntu test:

```
{"level":"info","ts":1605608771.9316554,"msg":"Listening to brim process pipe","fd":"brimfd"}
{"level":"info","ts":1605608771.9317698,"msg":"Starting","datadir":"/var/tmp/brimsec/itest/pcaps.test.ts/1/data/spaces","open_files_limit":65536,"pprof_routes":false,"suricata_supported":true,"zeek_supported":true}
{"level":"info","ts":1605608771.9320648,"msg":"Launching suricata updater"}
{"level":"info","ts":1605608771.9325843,"logger":"httpd","msg":"Listening","addr":"127.0.0.1:9867"}
{"level":"info","ts":1605608772.3187332,"msg":"Suricata updater completed"}
{"level":"error","ts":1605608772.3187568,"msg":"Running suricata updater","error":"/usr/lib/brim/resources/app/zdeps/suricata/suricataupdater exited with status 255: /usr/lib/brim/resources/app/zdeps/suricata/suricataupdater: line 8: /usr/lib/brim/resources/app/zdeps/suricata/update.yaml: Permission denied\n[3622] Failed to execute script suricata-update\nTraceback (most recent call last):\n  File \"suricata-update\", line 33, in <module>\n  File \"suricata/update/main.py\", line 1299, in main\n  File \"suricata/update/main.py\", line 1009, in _main\n  File \"suricata/update/config.py\", line 160, in init\nFileNotFoundError: [Errno 2] No such file or directory: '/usr/lib/brim/resources/app/zdeps/suricata/update.yaml'"}
```

**Comment by henridf** [comment]
@jameskerr thanks, makes sense! I hadn't realized the integration test was run from an installed package. Given that it is, we're running into #1216 here.

**Comment by henridf** [comment]
Windows 2019 failure looks like a library issue, which is surprising give that it passed previously on this branch. I'll investigate tomorrow.

**Comment by henridf** [comment]
As I couldn't repro the Windows failure (per https://github.com/brimsec/build-suricata/issues/31#issuecomment-730276022), I re-ran and now all tests pass. 🤷

**Comment by henridf** [comment]
@jameskerr I just merged master which now has an updated version of zq/suricata which should fix the windows issue that @philrz ran into above. Please consider this ready for final review!

**Review by jameskerr — APPROVED** [review]
I found 1 spelling error in the docs, but then this looks good to me!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -47,6 +47,12 @@ As an npm postinstall step, a [zeek artifact](https://github.com/brimsec/zeek/re
...
 
+### suricata
+
+Brim uses [Suricata](https://suricata-ids.org) to extract Suricata alert logs from packet captures. These alerts are also transformed into ZNG and combined with the zeek logs. Overall this feature's operation and configuration are similar to those of the Zeek feature. Here, the runner is passed to zqd with the `-suricatarunner` option, and it can be overridden via the Brim UI or the environment variable `BRIM_SURICATA_RUNNER`.
+
+The Suricata rules are updated every time Brim is launched. The update itself is started by zqd by invoking a suricata-update runner that is passed to it as `-suricataupdater`. The default updater uses the default Emerging Threads Open rules, like vanilla Suricata. A custom updater can be used; to do so change the updater path via the Brim UI or the environment variable `BRIM_SURICATA_UPDATER`.
```
[the remark this answers is not in the document]
Comment (henridf) [reply CONTRIBUTING.md]: oops

**Comment by henridf** [comment]
closes #1210

**Pull request merged**